PlatformIO build script hooks to wrong phase — transpile runs after compilation

## Description

`cnext --pio-install` generates a build script that hooks the transpile step to the `buildprog` target using `env.AddPreAction("buildprog", transpile_cnext)`. In PlatformIO, `buildprog` fires **after** compile + link (it creates the final firmware binary). This means .cnx files are transpiled too late — the compiler needs the generated .c/.cpp files to exist before compilation, not after.

## Reproduction

1. Run `cnext --pio-install` in a PlatformIO project
2. Delete any pre-existing generated .c/.cpp files
3. Run `pio run`

**Result:** Build fails at link time with `undefined reference to 'setup()'` / `undefined reference to 'loop()'`. The transpile step **never runs** because the build fails before reaching the `buildprog` target.

**Generated script (relevant line):**
```python
env.AddPreAction("buildprog", transpile_cnext)
```

**Build output proving wrong order:**
```
Compiling .pio/build/.../src/display_st7701.cpp.o    <-- compilation
Compiling .pio/build/.../src/i2c_driver.cpp.o
Compiling .pio/build/.../src/main.cpp.o
...
Linking .pio/build/.../firmware.elf                    <-- linking
Building .pio/build/.../firmware.bin                   <-- buildprog target
transpile_cnext(["buildprog"], ["...firmware.bin"])     <-- transpile runs HERE (too late!)
Transpiling 4 c-next files...
```

## Expected Behavior

The transpile step should run **before** compilation. In PlatformIO, this can be achieved by running the transpile at script import time (top-level in the `pre:` script) rather than as a pre-action on `buildprog`.

Example fix:
```python
Import("env")
import subprocess
from pathlib import Path

# Run transpile immediately at import time (before compilation)
src_dir = Path("src")
if src_dir.exists():
    cnx_files = list(src_dir.rglob("*.cnx"))
    if cnx_files:
        print(f"Transpiling {len(cnx_files)} c-next files...")
        for cnx_file in cnx_files:
            result = subprocess.run(
                ["cnext", str(cnx_file)],
                check=True, capture_output=True, text=True
            )
            print(f"  ✓ {cnx_file.name}")
```

## Secondary Issue

When the transpile does run (on subsequent builds where .cpp files already exist), it generates **new** .c files (if cppRequired config is broken, see #827). On the next build, PlatformIO compiles **both** the .c and .cpp files, causing duplicate symbol errors or C compilation failures for C++ APIs like `Wire` and `Serial`.

## Version

cnext v0.2.1

--- a/src/cli/PlatformIOCommand.ts
+++ b/src/cli/PlatformIOCommand.ts
@@ -44,11 +44,14 @@ class PlatformIOCommand {
     const { pioIniPath, scriptPath } = getPioProjectPaths();
 
     // Create cnext_build.py script
+    // Issue #833: Run transpilation at import time (before compilation),
+    // not as a pre-action on buildprog (which runs after compilation)
     const buildScript = `Import("env")
 import subprocess
+import sys
 from pathlib import Path
 
-def transpile_cnext(source, target, env):
+def transpile_cnext():
     """Transpile all .cnx files before build"""
     # Find all .cnx files in src directory
     src_dir = Path("src")
@@ -73,9 +76,10 @@ def transpile_cnext(source, target, env):
         except subprocess.CalledProcessError as e:
             print(f"  ✗ Error: {cnx_file.name}")
             print(e.stderr)
-            env.Exit(1)
+            sys.exit(1)
 
-env.AddPreAction("buildprog", transpile_cnext)
+# Run transpilation at import time (before compilation starts)
+transpile_cnext()
 `;
 
     writeFileSync(scriptPath, buildScript, "utf-8");

--- a/src/cli/__tests__/PlatformIOCommand.test.ts
+++ b/src/cli/__tests__/PlatformIOCommand.test.ts
@@ -68,6 +68,30 @@ describe("PlatformIOCommand", () => {
       expect(scriptCall?.[1]).toContain("transpile_cnext");
     });
 
+    it("generates script that runs at import time, not as buildprog pre-action (issue #833)", () => {
+      // Issue #833: buildprog fires AFTER compilation, so transpile runs too late
+      // The script should run transpilation at import time (before compilation)
+      vi.mocked(fs.existsSync).mockReturnValue(true);
+      vi.mocked(fs.readFileSync).mockReturnValue("[env:esp32]\n");
+
+      PlatformIOCommand.install();
+
+      const writeCalls = vi.mocked(fs.writeFileSync).mock.calls;
+      const scriptCall = writeCalls.find((call) =>
+        (call[0] as string).includes("cnext_build.py"),
+      );
+
+      expect(scriptCall).toBeDefined();
+      const scriptContent = scriptCall?.[1] as string;
+
+      // Should NOT use AddPreAction (runs too late)
+      expect(scriptContent).not.toContain("AddPreAction");
+      expect(scriptContent).not.toContain("buildprog");
+
+      // Should call transpile_cnext() at top level (import time)
+      expect(scriptContent).toContain("transpile_cnext()");
+    });
+
     it("adds extra_scripts to platformio.ini when not present", () => {
       vi.mocked(fs.existsSync).mockReturnValue(true);
       vi.mocked(fs.readFileSync).mockReturnValue(

